down.gemspec

```diff
@@ -20,7 +20,11 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency "minitest", "~> 5.8"
   spec.add_development_dependency "mocha", "~> 1.5"
   spec.add_development_dependency "rake"
-  spec.add_development_dependency "http", "~> 4.3"
+  if RUBY_VERSION >= "2.5"
+    spec.add_development_dependency "http", "~> 5.0"
+  else
+    spec.add_development_dependency "http", "~> 4.3"
+  end
```

Comment:
> OK thanks, I'm going to file a PR to their CHANGELOG adding that then.... seems important to be in changelog.
Thanks for that.
> 
> Not sure the right way to handle this over here, as per my comment I'm concerned about the way you are doing it, but maybe I'm wrong? Curious for maintainer opinion.

I'm answering your other comment but I need some time to order my ideas and do a proper explanation :smile:

Reply:
> This doesn't actually work to do what you want, I'm afraid.
> 
> Logic in a gemspec is executed at _build_ time, so which of these gets included will depend on what version of ruby the _releaser_ was using.
> 
> There isn't any way to do conditional logic in a gemspec like this that applies at runtime/execution time different for different apps.
> 
> However... I'm confused that this is only a _development_ dependency not an actual runtime dependency... _maybe_ this is ok? Is this really _only_ about CI? But it's very confusing/misleading, and will result in a gem released to rubygems whose `development_dependency` for http depends on what version of ruby the releaser was running...
> 
> I am not sure what to do, but I think ideally there is a different solution than this. Or at least some very explicit comments. Because this sort of thing is an error people make in gemspecs a lot, and 99% of the time it is an error, and causes big problems.

I will try to do my best to explain what lead me to this solution.

`down` has several backends ([readme](https://github.com/janko/down#backends)). You don't always need `http` gem to use `down`. If you use the default backend, there is no point in adding `http` to your system. This is by, as I understand, it's not a gemspec `dependency`.

It's needed as a `development_dependency` to test `Down::Http` backend. But also because some specs use it to build the expected result ([example](https://github.com/janko/down/blob/master/test/net_http_test.rb#L207)) and because `test_helper` uses it [to wait for a service](https://github.com/janko/down/blob/master/test/test_helper.rb#L56).

Since `http` drop ruby 2.3 and 2.4 support `down` could do the same. That's fair. But users running old rubies (2.3 and 2.4 have reached EOL) can still use `Down::Http` if they use `4.x` version. Drop old rubies support is not a requirement to use http 5.0, so I took the conservative alternative and let specs run with 4.x on old rubies (asserting that this works). 

Why the condition logic is in the gemspec and not in another place (Gemfile for example)? Because there are already some `development_dependencies` handled that way.